### PR TITLE
Feature/plans v2

### PR DESCRIPTION
**Pull request by khaykov**
This PR replaces existing Plans screen with new one, which uses Flux-C to fetch plan details, and looks visually similar to iOS.

This PR is a bit on a bigger side, but the majority of code used in it comes from History (which comes from ActivityLog). Also, I felt that for such a simple functionality dealing with feature branches/flags would be too much.
 
Functionality to test:

- Plans with features are loaded from the [endpoint](https://public-api.wordpress.com/wpcom/v2/plans/mobile) and displayed as-is, in the order they came.
- Plans are cached when you open the Plans screen for the first time (and with each subsequent request). If you visit Plans screen later without an internet connection, you will see cached Plans.
- If you visit Plans screen for the first time without an internet connection you will see an empty view.

Remaining functionality (detecting active plan, etc.) is TBD and will come later.

Design:

There is no official design, I mimicked design of Plans in iOS, with some minor chages.

plans_list

plan_details

plan_details_more

screenshot_1549513107

### CODE REVIEW

**Comment by iamthomasbishop** [comment]
@khaykovLooking good, nice work! 

One question: Where is the extra padding on each cell in the Plan Detail coming from? I'm seeing the usual 16dp gutters plus it looks like another bit of padding on each cellg, is this because you can swipe between plans? it doesn't look bad by any means, I'm just curiou how it's being defined.

The only thing I might  consider changing is to remove the dividers between cells, as it might clean the layout up a bit, but my feelings aren't strong on that matter. 😀

**Comment by khaykov** [comment]
@iamthomasbishop Thanks for taking a look! I reused a cell from another layout, and padding came from there :)

Without that padding and dividers layout looks like this:
no_padding_no_dividers

Without padding but with dividers:
screenshot_1549516000

**Comment by iamthomasbishop** [comment]
> Without that padding and dividers layout looks like this:

I kind of like it without dividers, but the margin between rows is too narrow in that case. If we were to bump the padding a bit between rows – maybe double what it is (16dp right now?) – I think we'd be in business.

**Comment by khaykov** [comment]
@iamthomasbishop Fixed! Now it looks like this:

screenshot_1549604415

File: WordPress/src/main/java/org/wordpress/android/ui/plans/PlansViewModel.kt [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+    @Subscribe(threadMode = ThreadMode.MAIN)
+    @SuppressWarnings("unused")
+    fun onPlanOffersFetched(event: OnPlanOffersFetched) {
+        if (event.isError) {
+            _listStatus.value = PlansListStatus.ERROR
+            AppLog.e(T.API, "An error occurred while fetching plans")
+        } else {
+            _listStatus.value = PlansListStatus.DONE
+        }
+        _plans.value = event.planOffers // in case of PlansListStatus.ERROR this might contain cached plans
```
Comment (theck13) [inline WordPress/src/main/java/org/wordpress/android/ui/plans/PlansViewModel.kt]: When we use cached plans, should we show a message to the user so they know these plans may not be completely up-to-date?  That may be a question for @iamthomasbishop to consider too.

**Comment by iamthomasbishop** [comment]
@khaykov That looks a lot better, but still might be a little tight. I think if we add 8pt more top and bottom margin between cells, that'd be good.

I also agree w/ Tyler – we might want to change the title to `grey_dark`. 

> When we use cached plans, should we show a message to the user so they know these plans may not be completely up-to-date? That may be a question for @iamthomasbishop to consider too.

@theck13 How/when are we caching the plans? More specifically, how out-of-date are they likely to get? I imagine if they're a day or less out-of-date, the likelihood of an issue is low. However, if they're only getting cached less regularly, it could be an issue. Is there a way to determine how long it has been since last cache or if there has been changes made – and only show a message in that case? I would just prefer to show a message until there's a reason to.

**Comment by theck13** [comment]
> How/when are we caching the plans? More specifically, how out-of-date are they likely to get?

The cache is updated every time the plans are viewed.  That means they could be very out-of-date if the user hasn't viewed the plans in a while.

> Is there a way to determine how long it has been since last cache

Yes, we can get the time whenever the plans are viewed and compare it to the previous time the plans were viewed to determine how long it has been.

> Is there a way to determine... if there has been changes made

Ideally, we would only show a message if the cached plans are different than the current plans, but that requires a network connection to retrieve the current plans and compare them for differences.

**Comment by khaykov** [comment]
@iamthomasbishop Added more padding :)

screenshot_1549659741

One thing we can do with outdated plans is utilize "NO CONNECTION" bar with different copy. Clicking on it should refresh plans and hide the message if sucess. I wonder what copy will fit there?
[Image from Gyazo](https://gyazo.com/a9f8a5c3b022b8118fd4587fd55b4920)

**Comment by iamthomasbishop** [comment]
Sorry for all of the back and forth! Between the cells looks perfect (24dp?), but I think the top of the page is a little unbalanced with that now. I'm thinking 24dp between elements across the board feels most balanced:

image

**Comment by khaykov** [comment]
@iamthomasbishop Thanks for exact dimensions. Now it should be uniform 24dp :)
screenshot_1549668389

**Comment by theck13** [comment]
> One thing we can do with outdated plans is repurpose "NO CONNECTION" bar with different copy.

That sounds good to me.  It's an interface component we can reuse and users should be familiar with it from other parts of the app when there is no connectivity.

**Comment by iamthomasbishop** [comment]
Looking good @khaykov thanks! 

I agree w/ Tyler, let's re-use that connection message component. 

@theck13 Would it be doable for out-of-date plans to show the messaging as "Plans may be out-of-date. Reload to update." and if there is no connection, then show the alert above or some variation of it pending copy review?

**Comment by theck13** [comment]
> Would it be doable for out-of-date plans to show the messaging as "Plans may be out-of-date. Reload to update."

Determining whether the plans are out-of-date is the hard part.  If there is a network connection, the plans are updated and no out-of-date check is needed.  If there is no network connection, we have no way of knowing if plans are out-of-date other than the amount of time since the plans were last updated in the app.  The greater the amount of time since the plans were last accessed increases the likelihood the plans are out-of-date, but it is not a certainty.

What about _not_ showing the cached plans by default, displaying the empty view when there is no connection, and giving the option to view cached plans to the user?  The empty view description will give us more text component to explain what's happening than the yellow connection status.  That would help ensure the user is aware the plans _may_ be out-of-date, but they can still view some details while offline.  I don't think "cached" is the best term, but I was thinking something like this.

**Comment by loremattei** [comment]
Hey! I'm moving this one to 11.9 since 11.8 has been cut. If it has to land 11.8, please feel free to move it back, target the release branch and ping me to build a new beta.

**Comment by iamthomasbishop** [comment]
I think this would work, although as you mentioned the copy might need some tweaking. This would be my suggestion, but let's get copy review from Editorial:

Title: `No connection`
Body: `An internet connection is required to view Plans, so details and pricing may be out-of-date.`
Button: `View Plans Anyway`

**Comment by frosty** [comment]
> Body: `An internet connection is required to view Plans, so details and pricing may be out-of-date.`

We don't currently make any mention of pricing in the app, do we?

**Comment by iamthomasbishop** [comment]
> We don't currently make any mention of pricing in the app, do we?

That's right, my bad. We'd remove the reference to pricing.

Body: `An internet connection is required to view Plans, so details may be out-of-date.`

**Comment by khaykov** [comment]
@iamthomasbishop @theck13 Thank you so much for great suggestions! 🙇
I made a couple of changes to AEV's to handle various states of errors/cache.
From left to right:

- Not an error, but no plans came from endpoint (could potentially happen).
- Network error and we have cache.
- Network error but we don't have cache (eg, when user navigates to the Plan's screen for the first time without internet conenction).

aevs

In addition, I added a change where if user has Plan details open, leaves app and Plans acitivty is killed by OS the Plan details will be dismissed when he comes back and we will try to fetch fresh plans. (Hope this makes sense).

File: WordPress/src/main/res/layout/plan_details_fragment.xml [hunk 2]
```diff
@@ -1,64 +1,68 @@
...
             android:id="@+id/image_plan_icon"
             android:layout_width="@dimen/plan_icon_size"
             android:layout_height="@dimen/plan_icon_size"
             android:layout_gravity="center_horizontal"
-            android:layout_marginTop="@dimen/margin_extra_large"
-            tools:background="@drawable/penandink_223dp"
-            android:contentDescription="@null"/>
+            android:contentDescription="@null"
+            android:src="@drawable/ic_plans_white_24dp"
+            android:tint="@color/grey"/>
```
Comment (theck13) [inline WordPress/src/main/res/layout/plan_details_fragment.xml]:
```suggestion
            tools:tint="@color/grey"/>
```

The `android:tint` attribute causes the plan icon to be `grey` (`#87a6bc`) even after it has been loaded form the icon URL.  See the screenshot below for illustration.



We should change the namespace to `tools`.

File: WordPress/src/main/res/layout/plans_list_item.xml [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+        <ImageView
+            android:id="@+id/plan_image"
+            android:layout_width="@dimen/activity_log_icon_size"
+            android:layout_height="@dimen/activity_log_icon_size"
+            android:layout_alignParentStart="true"
+            android:layout_centerVertical="true"
+            android:layout_margin="@dimen/activity_log_icon_margin"
+            android:contentDescription="@string/description_user"
+            android:src="@drawable/ic_plans_white_24dp"
+            android:tint="@color/grey">
```
Comment (theck13) [inline WordPress/src/main/res/layout/plans_list_item.xml]:
```suggestion
            tools:tint="@color/grey"/>
```

The `android:tint` attribute causes the plan icons to be `grey` (`#87a6bc`) even after they have been loaded form the icon URL.  See the screenshot below for illustration.



We should change the namespace to `tools`.

**Comment by khaykov** [comment]
@theck13 Good catch! No idea how I managed to miss it 🤦‍♂️

**Comment by iamthomasbishop** [comment]
All looks good to me! :shipit:

**Pull request merged**